src/Lecture2.hs

```diff
@@ -70,7 +73,23 @@ return the removed element.
 >>> removeAt 10 [1 .. 5]
 (Nothing,[1,2,3,4,5])
 -}
-removeAt = error "TODO"
+{- This is ugly and super slow but couldn't do better.
+I asked you for an error resolve of my another prettier solution,
+and it works fine but it needs pragma ScopedTypeVariables and "forall a ."
+that was not in the lectures yet. So I keep this one until then.
+-}
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt _ [] = (Nothing, [])
+removeAt n lst 
+    | n < 0 = (Nothing, lst)
+    | otherwise = 
+        let newlst = []
+        in go n lst newlst
+            where
+                go :: Int -> [a] -> [a] -> (Maybe a, [a])
+                go _ [] newlst = (Nothing, newlst)
+                go 0 (x : xs) newlst = (Just x, newlst ++ xs)
+                go n (x : xs) newlst = go (n - 1) (xs) (newlst ++ [x])
```

Comment:
That's a correct implementation 👍🏻 

Alternatively, you can use a standard function [splitAt](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:splitAt) instead of writing manual recursion 🔃

If you want to keep a recursive solution, you can do improve the performance by avoiding to have a separate `newlst` argument. You can write the function like this:

```haskell
go n (x : xs) =
    let (val, rest) = go (n - 1) xs
    in (val, x : rest)
```

This way you'll avoid adding elements to the end of the list with `++` which is slow 🐌

Reply:
Thanks! I've seen "splitAt" but I thought I could try to implement it myself 😄